### DISCUSSION TITLE
API monetization plugin

### DISCUSSION THREAD

**Discussion by robertlagrant** [Kong/kong#7313] [Ideas and feature requests]
# Summary

API monetization allows usage tracking and costing for internal and external consumers. Adding this to Kong could provide per-consumer API bills, which could pave the way for Stripe or equivalent plugins to provide payments, cost management plugins and other useful features.

**Comment by quenbyako** [comment]
Kong is about authentication and proxying, not about billing. But, i really need manage billing in my projects too, so i think, we need some adapter  for killbill, for example. Killbill is good billing system i think.

**Comment by jeremyjpj0916** [comment]
Ive always thought a neat v1 concept of a billing plugin could be tracking all Kong consumer resources and their monthly successful tx consumption by 2XX status codes and billing based on that in a daily/weekly/monthly email of text or something as a first shot. Then you have a very basic way to bill customers directly with Kong(then you write 3rd party apps to parse and handle that email data if you want more automation). Or maybe the plugin API itself exposes the data so it can be programmatically queried upon demand(usage for the last X days/weeks/months by status). To optimize its storage into the db don't bombard with tx's but take tx rates in chunks and + it into the db periodically, every few mins or so as to keep db level stress to a minimum, configurable of course :).  Maybe I will write a plugin like this if I get free time for fun stuff, also my knowledge for writing to the kong db outside of schema.lua changes is weak so I gotta dip my toes there heh.

**Comment by quenbyako** [comment]
@jeremyjpj0916 i don't understand actually, how to write plugins for kong, but:

1) i working on killbill rewriting on go, but strictly based on kong features (curently it's private repo, but i'll open it soon). if you want to make plugin for kong, please, contact me [telegram](https://t.me/zhopakotabot) or [mail](mailto:rcooper.xelaj@protonmail.com), if you'll have "free time for fun stuff", i really want to help with this feature, but i can't do it in just 2 hands.

2) i working on [free kong UI](https://github.com/xelaj/kong_abu), it can be helpful for management billing for consumers, and could be add billing functionality in this ui

**Reply by AliElMoussawi** [reply]
Hi ,
I have a question about kill Bill
According to the documentation, Kill Bill provides the interactions of the management layer by APIs and integrations. I want to know more about the Kill Bill customer layer that allow customer account to interaction with Kill Bill APIs, if there any available plugin that provide this service or any mechanism you may suggest that will be helpful. 

in Kill Bill user Account --Kill Bill APIs-->Kill Bill ->customer account
I want user user (customer account)--Kill Bill APIs--> Kill Bill(with limitations)
